dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/MetricInstrumentor.java

```
@@ -235,112 +237,223 @@ private boolean isCompatible(Type argType, Type expectedType) {
+      return null;
+    }
+
+    @Override
+    public VisitorResult visit(StartsWithExpression startsWithExpression) {
+      return null;
+    }
+
+    @Override
+    public VisitorResult visit(SubStringExpression subStringExpression) {
+      return null;
+    }
+
+    @Override
+    public VisitorResult visit(ValueRefExpression valueRefExpression) {
+      String name = valueRefExpression.getSymbolName();
       InsnList insnList = new InsnList();
-      Type currentType = tryRetrieve(name, insnList);
-      if (currentType == null) {
-        reportError("Cannot resolve symbol " + name);
-        return null;
+      Type currentType;
+      if (name.equals(ValueReferences.THIS)) {
+        currentType = Type.getObjectType(instrumentor.classNode.name);
+        insnList.add(new VarInsnNode(Opcodes.ALOAD, 0));
+        // stack [this]
+      } else {
+        currentType = tryRetrieve(name, insnList);
+        if (currentType == null) {
+          throw new InvalidValueException("Cannot resolve symbol " + name);
+        }
+        // stack [arg|local|field]
       }
-      convertIfRequired(currentType, expectedType, insnList);
-      return new ResolverResult(currentType, insnList);
+      return new VisitorResult(currentType, insnList);
     }
 
     @Override
-    public Object getMember(Object target, String name) {
-      Type currentType = null;
-      InsnList insnList = null;
-      if (target == Values.THIS_OBJECT) {
-        insnList = new InsnList();
-        currentType = tryRetrieveField(name, insnList);
-        convertIfRequired(currentType, expectedType, insnList);
-      } else if (target instanceof ResolverResult) {
-        ResolverResult result = (ResolverResult) target;
-        insnList = result.insnList;
-        currentType = followReferences(result.type, name, insnList);
+    public VisitorResult visit(GetMemberExpression getMemberExpression) {
+      VisitorResult result = getMemberExpression.getTarget().accept(this);
```

Results can be null - should we throw unsupported expression instead of nulls? that would make better error message and also avoid null checks. WDYT?

---

done